server/src/main/java/org/elasticsearch/transport/AbstractResultDeduplicator.java

```diff
@@ -28,9 +28,11 @@
 import java.util.function.BiConsumer;
 
 /**
- * Deduplicator that keeps track of requests that should not be sent/executed in parallel.
+ * Deduplicator for arbitrary keys and results that can be used to ensure a given action is only executed once at a time for a given
+ * request.
+ * @param <T> Request type
  */
-public final class TransportRequestDeduplicator<T> {
+public final class AbstractResultDeduplicator<T, R> {
```

Comment:
nit: Maybe we should name it just `ResultDeduplicator`? as we don't expect to extend this class.
Also, now that we're using it outside of the transport package, maybe it makes sense to move it to another package?

Reply:
++ to both, lets renamed and moved to the `.action` package :)